src/com/amazon/ion/impl/IonCursorBinary.java

```diff
@@ -178,8 +262,362 @@ class IonCursorBinary implements IonCursor {
+    }
+
+    /**
+     * Read the next byte from the stream, ensuring the byte is buffered.
+     * @return the byte, or -1 if the end of the stream has been reached.
+     */
+    private int slowReadByte() {
+        if (refillableState.isSkippingCurrentValue) {
+            // If the value is being skipped, the byte will not have been buffered.
+            return readByteWithoutBuffering();
+        }
+        if (!fillAt(peekIndex, 1)) {
+            return -1;
+        }
+        return slowPeekByte();
+    }
+
+    /**
+     * Shift all active container end indices left by the given amount. This is used when bytes have been shifted
+     * to the start of the buffer in order to make room at the end.
+     * @param shiftAmount the amount to shift left.
+     */
+    private void shiftContainerEnds(long shiftAmount) {
+        for (int i = containerIndex; i >= 0; i--) {
+            if (containerStack[i].endIndex > 0) {
+                containerStack[i].endIndex -= shiftAmount;
+            }
+        }
+    }
+
+    /**
+     * Shift all indices left by the given amount. This is used when data is moved in the underlying
+     * buffer either due to buffer growth or NOP padding being reclaimed to make room for a value that would otherwise
+     * exceed the buffer's maximum size.
+     * @param shiftAmount the amount to shift left.
+     */
+    private void shiftIndicesLeft(int shiftAmount) {
+        peekIndex = Math.max(peekIndex - shiftAmount, 0);
+        valuePreHeaderIndex -= shiftAmount;
+        valueMarker.startIndex -= shiftAmount;
+        valueMarker.endIndex -= shiftAmount;
+        checkpoint -= shiftAmount;
+        if (annotationSequenceMarker.startIndex > -1) {
+            annotationSequenceMarker.startIndex -= shiftAmount;
+            annotationSequenceMarker.endIndex -= shiftAmount;
+        }
+        shiftContainerEnds(shiftAmount);
+        refillableState.totalDiscardedBytes += shiftAmount;
     }
 
+    /**
+     * Fills the buffer with up to the requested number of additional bytes. It is the caller's responsibility to
+     * ensure that there is space in the buffer.
+     * @param numberOfBytesToFill the number of additional bytes to attempt to add to the buffer.
+     */
+    private void refill(long numberOfBytesToFill) {
+        int numberOfBytesFilled = -1;
+        try {
+            numberOfBytesFilled = refillableState.inputStream.read(buffer, (int) limit, (int) numberOfBytesToFill);
+        } catch (IOException e) {
+            throwAsIonException(e);
+        }
+        if (numberOfBytesFilled < 0) {
+            return;
+        }
+        limit += numberOfBytesFilled;
+    }
+
+    /**
+     * Seeks forward in the stream up to the requested number of bytes, from `offset`.
+     * @param numberOfBytes the number of bytes to seek from `offset`.
+     * @return true if the seek is complete; otherwise, false.
+     */
+    private boolean slowSeek(long numberOfBytes) {
```

Comment:
what would a fastSeek do (or not do) that this does?

Reply:
fastSeek would simply move `peekIndex` ahead to the endIndex of the value. This is performed inline (one line) where necessary in the non-slow methods.